### ISSUE TITLE
odbc.js throwing "Cannot call method 'close' of undefined"

### ISSUE THREAD

**Issue by jcwren** [w1nk/node-odbc#69]
I'm using the code below as a test case for node-odbc, and it works, up until the close() call, when it emits the following error:

```
[ { stage_no: 7,
    stage_name: 'Nag Nag Nag',
    classifier: 'No',
    classifier_no: '',
    scoring_type: 'Comstock',
    pepper_poppers_5: 16,
    pepper_poppers_10: 0,
    plates: 0,
    maximum_points: 80,
    times_run: 1,
    stage_deleted: 'No' } ] 
[]
Database connection closed

C:\jcw\nodeodbc\node_modules\odbc\lib\odbc.js:163
    self.conn.close(function (err) {
              ^
TypeError: Cannot call method 'close' of undefined
    at C:\jcw\nodeodbc\node_modules\odbc\lib\odbc.js:163:15
    at SimpleQueue.next (C:\jcw\nodeodbc\node_modules\odbc\lib\simple-queue.js:34:5)
    at SimpleQueue.maybeNext (C:\jcw\nodeodbc\node_modules\odbc\lib\simple-queue.js:22:10)
    at C:\jcw\nodeodbc\node_modules\odbc\lib\simple-queue.js:37:12
    at C:\jcw\nodeodbc\node_modules\odbc\lib\odbc.js:168:14
```

This is with node-odbc 0.6.10, nodejs 0.10.29, Windows 7/32

I'm not sure if it's a _correct_ solution, but adding a 'if (self.conn)' before the 'self.conn.close' at line 163 fixes it.

The test code is:

```
var Database = require ("odbc").Database;

var cn = 'DRIVER={Adaptive Server Anywhere 9.0};START=C:\\Program Files (x86)\\EzWinScore\\rteng9.exe;UID=dba;PWD=sql;DatabaseName=DBA;DatabaseFile=C:\\USPSA\\RBGC_2014-05-24.db;AutoStop=YES';

var db = new Database ();

db.open (cn, function (err) {
  if (err)
    return console.log (err);

  db.query ('SELECT * FROM "stage profiles" ORDER BY stage_no', function (err, data) {
    console.log (data);

    db.close (function () {
      console.log ("Database connection closed");
    });
  });
});
```

**Comment by wankdanker** [comment]
Wow. That's weird. I can't replicate that on my linux box. Will have to try on Windows when I get a chance hopefully on Thursday.

The pattern in your example is pretty much what I use in many projects and I've never hit this bug.

In your test case, try adding `console.log(db.conn)` before calling `db.close (...)`. Based on the error message I would expect it to log 'undefined'.

**Comment by wankdanker** [comment]
Correction, I have hit that bug. I just glazed right over it. Can't reproduce it on demand though. I've put some debug logging in place where I am randomly hitting it and will hopefully have an answer for you soon.

**Comment by ndelitski** [comment]
Have the same issue with the code below, hope this helps

``` coffee
query = require("./lib").execSql
query "testDb", """
        backup database "#{originalDbName}"
          to disk='C:\\tmp\\#{originalDbName}.bak'
          with init;
        restore database "#{copyDbName}"
          from disk='C:\\tmp\\#{originalDbName}.bak'
          with
            move '#{originalDbName}_Data' to 'C:\\data\\#{copyDbName}.mdf',
            move '#{originalDbName}_log' to  'C:\\data\\#{copyDbName}.ldf'
    """
```

`query()` method:

``` js
require('shelljs/global');

var conf = require('./../conf'),
    db = new (require('odbc').Database)({connectTimeout:20, loginTimeout:20});

exports.execSql = function (connStringName, sql, params, successCallback) {
    if (arguments.length==3) {
        successCallback=params;
        params=[];
    }

    db.open(conf.connectionStrings[connStringName], function(err) {
        if (err) {
            echo(err);
            exit(1);
        }
        db.query(sql, params, function(err, rows) {
            if (err) {
                echo(err);
                exit(1);
            }

            db.close();

            if (successCallback && typeof successCallback == 'function') 
                successCallback(rows);
        })
    });
}

```

**Comment by wankdanker** [comment]
Hey @ndelitski,

My guess is that in your case you may be experiencing something slightly different. The query you are executing may be creating multiple result sets since there are actually two commands in there, backup and restore.

The callback function signature for db.query is actually `function (err, rows, moreResultSets)` so you might want to try something like:

``` js
db.query(..., function (err, rows, moreResultSets) {
  if (!moreResultSets) {
    db.close();
    //also call your callback here
  }
})
```

**Comment by nevercast** [comment]
This occurred for me also when I was running queries from the callback of another query. One query would run, fire the callback, which would run another. At the completion of the second query it would close the database and control would be returned to the first query, which would then try to close the connection also. Because it was already closed this through the same error you're experiencing. 

I could have shuffled around the invoke order but instead I decided to attach a 'runningQueries' property to the database instance and only close when it was 0. Incrementing on db.query and decrementing on it's callback.

That's just my experience with it.
Anyway! The cause is, as @wankdanker said, closing an already closed connection. I believe this is a user code error more than a node-odbc error. Just check where you're closing :)

**Comment by wankdanker** [comment]
I believe this was fixed in a560095. Thanks for your help, @nevercast.
